README.md
NoriV2
====

[![Build Status](https://secure.travis-ci.org/savonrb/nori_v2.png)](http://travis-ci.org/savonrb/nori_v2)
[![Gem Version](https://badge.fury.io/rb/nori_v2.png)](http://badge.fury.io/rb/nori_v2)
[![Code Climate](https://codeclimate.com/github/savonrb/nori_v2.png)](https://codeclimate.com/github/savonrb/nori_v2)
[![Coverage Status](https://coveralls.io/repos/savonrb/nori_v2/badge.png?branch=master)](https://coveralls.io/r/savonrb/nori_v2)


Really simple XML parsing ripped from Crack which ripped it from Merb.  
NoriV2 was created to bypass the stale development of Crack, improve its XML parser  
and fix certain issues.

``` ruby
parser = NoriV2.new
parser.parse("<tag>This is the contents</tag>")
# => { 'tag' => 'This is the contents' }
```

NoriV2 supports pluggable parsers and ships with both REXML and Nokogiri implementations.  
It defaults to Nokogiri since v2.0.0, but you can change it to use REXML via:

``` ruby
NoriV2.new(:parser => :rexml)  # or :nokogiri
```

Make sure Nokogiri is in your LOAD_PATH when parsing XML, because NoriV2 tries to load it  
when it's needed.


Typecasting
-----------

Besides regular typecasting, NoriV2 features somewhat "advanced" typecasting:

* "true" and "false" String values are converted to `TrueClass` and `FalseClass`.
* String values matching xs:time, xs:date and xs:dateTime are converted
  to `Time`, `Date` and `DateTime` objects.

You can disable this feature via:

``` ruby
NoriV2.new(:advanced_typecasting => false)
```


Namespaces
----------

NoriV2 can strip the namespaces from your XML tags. This feature might raise  
problems and is therefore disabled by default. Enable it via:

``` ruby
NoriV2.new(:strip_namespaces => true)
```


XML tags -> Hash keys
---------------------

NoriV2 lets you specify a custom formula to convert XML tags to Hash keys.  
Let me give you an example:

``` ruby
parser = NoriV2.new(:convert_tags_to => lambda { |tag| tag.snakecase.to_sym })

xml = '<userResponse><accountStatus>active</accountStatus></userResponse>'
parser.parse(xml)  # => { :user_response => { :account_status => "active" }
```

Dashes and underscores
----------------------

NoriV2 will automatically convert dashes in tag names to underscores.
For example:

```ruby
parser = NoriV2.new
parser.parse('<any-tag>foo bar</any-tag>')  # => { "any_tag" => "foo bar" }
```

You can control this behavior with the `:convert_dashes_to_underscores` option:

```ruby
parser = NoriV2.new(:convert_dashes_to_underscores => false)
parser.parse('<any-tag>foo bar</any-tag>') # => { "any-tag" => "foo bar" }
```


CHANGELOG.md
# 2.4.0 (2014-04-19)

* Change: Dropped support for ruby 1.8, rubinius and ree

* Feature: Added `:convert_attributes` feature similar to `:convert_tags_to`

* Feature: Added `:convert_dashes_to_underscore` option

# 2.3.0 (2013-07-26)

* Change: `NoriV2#find` now ignores namespace prefixes in Hash keys it is searching through.

* Fix: Limited Nokogiri to < 1.6, because v1.6 dropped support for Ruby 1.8.

# 2.2.0 (2013-04-25)

* Feature: [#42](https://github.com/savonrb/nori_v2/pull/42) adds the `:delete_namespace_attributes`
  option to remove namespace attributes like `xmlns:*` or `xsi:*`.

# 2.1.0 (2013-04-21)

* Feature: Added `NoriV2.hash_key` and `NoriV2#find` to work with Hash keys generated by NoriV2.
  Original issue: [savonrb/savon#393](https://github.com/savonrb/savon/pull/393)

# 2.0.4 (2013-02-26)

* Fix: [#37](https://github.com/savonrb/nori_v2/issues/37) special characters
  problem on Ruby 1.9.3-p392.

# 2.0.3 (2013-01-10)

* Fix for remote code execution bug. For more in-depth information, read about the
  recent [Rails hotfix](https://groups.google.com/forum/?fromgroups=#!topic/rubyonrails-security/61bkgvnSGTQ).
  Please make sure to upgrade now!

# 2.0.2 (YANKED)

* Yanked because of a problem with XML that starts with an instruction tag.

# 2.0.1 (YANKED)

* Yanked because of a problem with XML that starts with an instruction tag.

# 2.0.0 (2012-12-12)

Please make sure to read the updated README for how to use the new version.

* Change: NoriV2 now defaults to use the Nokogiri parser.

* Refactoring: Changed the `NoriV2` module to a class. This might cause problems if you
  included the `NoriV2` module somewhere in your application. This use case was removed
  for overall simplicity.

* Refactoring: Changed the interface to remove any global state. The global configuration
  is gone and replaced with simple options to be passed to `NoriV2.new`.

    ``` ruby
    parser = NoriV2.new(strip_namespaces: true)
    parser.parse(xml)
    ```

* Refactoring: Removed the `NoriV2::Parser` module methods. After refactoring the rest,
  there was only a single method left for this module and that was moved to `NoriV2`.

* Fix: [#16](https://github.com/savonrb/nori_v2/issues/16) strip XML passed to NoriV2.

## 1.1.5 (2013-03-03)

* Fix: [#37](https://github.com/savonrb/nori_v2/issues/37) special characters
  problem on Ruby 1.9.3-p392.

## 1.1.4 (2013-01-10)

* Fix for remote code execution bug. For more in-depth information, read about the
  recent [Rails hotfix](https://groups.google.com/forum/?fromgroups=#!topic/rubyonrails-security/61bkgvnSGTQ).
  Please make sure to upgrade now!

## 1.1.3 (2012-07-12)

* Fix: Merged [pull request 21](https://github.com/savonrb/nori_v2/pull/21) to fix an
  issue with date/time/datetime regexes not matching positive time zone offsets and
  datetime strings with seconds.

## 1.1.2 (2012-06-30)

* Fix: Reverted `Object#xml_attributes` feature which is planned for version 2.0.

## 1.1.1 (2012-06-29) - yanked

* Fix: Merged [pull request 17](https://github.com/savonrb/nori_v2/pull/17) for improved
  xs:time/xs:date/xs:dateTime regular expression matchers.

## 1.1.0 (2012-02-17)

* Improvement: Merged [pull request 9](https://github.com/savonrb/nori_v2/pull/9) to
  allow multiple configurations of NoriV2.

* Fix: Merged [pull request 10](https://github.com/savonrb/nori_v2/pull/10) to handle
  date/time parsing errors. Fixes a couple of similar error reports.

## 1.0.2 (2011-07-04)

* Fix: When specifying a custom formula to convert tags, XML attributes were ignored.
  Now, a formula is applied to both XML tags and attributes.

## 1.0.1 (2011-06-21)

* Fix: Make sure to always load both StringWithAttributes and StringIOFile
  to prevent NameError's.

## 1.0.0 (2011-06-20)

* Notice: As of v1.0.0, NoriV2 will follow [Semantic Versioning](http://semver.org).

* Feature: Added somewhat advanced typecasting:

  What this means:

  * "true" and "false" are converted to TrueClass and FalseClass
  * Strings matching an xs:time, xs:date and xs:dateTime are converted
    to Time, Date and DateTime objects.

  You can disable this feature via:

      NoriV2.advanced_typecasting = false

* Feature: Added an option to strip the namespaces from every tag.
  This feature might raise problems and is therefore disabled by default.

      NoriV2.strip_namespaces = true

* Feature: Added an option to specify a custom formula to convert tags.
  Here's an example:

      NoriV2.configure do |config|
        config.convert_tags_to { |tag| tag.snake_case.to_sym }
      end

      xml = '<userResponse><accountStatus>active</accountStatus></userResponse>'
      parse(xml).should ## { :user_response => { :account_status => "active" }

## 0.2.4 (2011-06-21)

* Fix: backported fixes from v1.0.1

## 0.2.3 (2011-05-26)

* Fix: Use extended core classes StringWithAttributes and StringIOFile instead of
  creating singletons to prevent serialization problems.

## 0.2.2 (2011-05-16)

* Fix: namespaced xs:nil values should be nil objects.

## 0.2.1 (2011-05-15)

* Fix: Changed XML attributes converted to Hash keys to be prefixed with an @-sign.
  This avoids problems with attributes and child nodes having the same name.

      <multiRef id="id1">
        <approved xsi:type="xsd:boolean">true</approved>
        <id xsi:type="xsd:long">76737</id>
      </multiRef>

  is now translated to:

      { "multiRef" => { "@id" => "id1", "id" => "76737", "approved" => "true" } }

## 0.2.0 (2011-04-30)

* Removed JSON from the original Crack basis
* Fixed a problem with Object#blank?
* Added swappable parsers
* Added a Nokogiri parser with you can switch to via:

      NoriV2.parser = :nokogiri

## 0.1.7 2010-02-19
* 1 minor patch
  * Added patch from @purp for ISO 8601 date/time format

## 0.1.6 2010-01-31
* 1 minor patch
  * Added Crack::VERSION constant - http://weblog.rubyonrails.org/2009/9/1/gem-packaging-best-practices

## 0.1.5 2010-01-27
* 1 minor patch

lib/nori_v2/xml_utility_node.rb
require "date"
require "time"
require "yaml"
require "bigdecimal"

require "nori_v2/string_with_attributes"
require "nori_v2/string_io_file"

class NoriV2

  # This is a slighly modified version of the XMLUtilityNode from
  # http://merb.devjavu.com/projects/merb/ticket/95 (has.sox@gmail.com)
  #
  # John Nunemaker:
  # It's mainly just adding vowels, as I ht cd wth n vwls :)
  # This represents the hard part of the work, all I did was change the
  # underlying parser.
  class XMLUtilityNode

    # Simple xs:time Regexp.
    # Valid xs:time formats
    # 13:20:00          1:20 PM
    # 13:20:30.5555     1:20 PM and 30.5555 seconds
    # 13:20:00-05:00    1:20 PM, US Eastern Standard Time
    # 13:20:00+02:00    1:20 PM, Central European Standard Time
    # 13:20:00Z         1:20 PM, Coordinated Universal Time (UTC)
    # 00:00:00          midnight
    # 24:00:00          midnight

    XS_TIME = /^\d{2}:\d{2}:\d{2}[Z\.\-\+]?\d*:?\d*$/

    # Simple xs:date Regexp.
    # Valid xs:date formats
    # 2004-04-12           April 12, 2004
    # -0045-01-01          January 1, 45 BC
    # 12004-04-12          April 12, 12004
    # 2004-04-12-05:00     April 12, 2004, US Eastern Standard Time, which is 5 hours behind Coordinated Universal Time (UTC)
    # 2004-04-12+02:00     April 12, 2004, Central European Summer Time, which is 2 hours ahead of Coordinated Universal Time (UTC)
    # 2004-04-12Z          April 12, 2004, Coordinated Universal Time (UTC)

    XS_DATE = /^[-]?\d{4}-\d{2}-\d{2}[Z\-\+]?\d*:?\d*$/

    # Simple xs:dateTime Regexp.
    # Valid xs:dateTime formats
    # 2004-04-12T13:20:00           1:20 pm on April 12, 2004
    # 2004-04-12T13:20:15.5         1:20 pm and 15.5 seconds on April 12, 2004
    # 2004-04-12T13:20:00-05:00     1:20 pm on April 12, 2004, US Eastern Standard Time
    # 2004-04-12T13:20:00+02:00     1:20 pm on April 12, 2004, Central European Summer Time
    # 2004-04-12T13:20:15.5-05:00   1:20 pm and 15.5 seconds on April 12, 2004, US Eastern Standard Time
    # 2004-04-12T13:20:00Z          1:20 pm on April 12, 2004, Coordinated Universal Time (UTC)

    XS_DATE_TIME = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}[\.Z]?\d*[\-\+]?\d*:?\d*$/

    def self.typecasts
      @@typecasts
    end

    def self.typecasts=(obj)
      @@typecasts = obj
    end

    def self.available_typecasts
      @@available_typecasts
    end

    def self.available_typecasts=(obj)
      @@available_typecasts = obj
    end

    self.typecasts = {}
    self.typecasts["integer"]       = lambda { |v| v.nil? ? nil : v.to_i }
    self.typecasts["boolean"]       = lambda { |v| v.nil? ? nil : (v.strip != "false") }
    self.typecasts["datetime"]      = lambda { |v| v.nil? ? nil : Time.parse(v).utc }
    self.typecasts["date"]          = lambda { |v| v.nil? ? nil : Date.parse(v) }
    self.typecasts["dateTime"]      = lambda { |v| v.nil? ? nil : Time.parse(v).utc }
    self.typecasts["decimal"]       = lambda { |v| v.nil? ? nil : BigDecimal(v.to_s) }
    self.typecasts["double"]        = lambda { |v| v.nil? ? nil : v.to_f }
    self.typecasts["float"]         = lambda { |v| v.nil? ? nil : v.to_f }
    self.typecasts["string"]        = lambda { |v| v.to_s }
    self.typecasts["base64Binary"]  = lambda { |v| v.unpack('m').first }

    self.available_typecasts = self.typecasts.keys

    def initialize(options, name, attributes = {})
      @options = options
      @name = NoriV2.hash_key(name, options)

      if converter = options[:convert_attributes_to]
        intermediate = attributes.map {|k, v| converter.call(k, v) }.flatten
        attributes = Hash[*intermediate]
      end

      # leave the type alone if we don't know what it is
      @type = self.class.available_typecasts.include?(attributes["type"]) ? attributes.delete("type") : attributes["type"]

      @nil_element = false
      attributes.keys.each do |key|
        if result = /^((.*):)?nil$/.match(key)
          @nil_element = attributes.delete(key) == "true"
          attributes.delete("xmlns:#{result[2]}") if result[1]
        end
        attributes.delete(key) if @options[:delete_namespace_attributes] && key[/^(xmlns|xsi)/]
      end
      @attributes = undasherize_keys(attributes)
      @children = []
      @text = false
    end

    attr_accessor :name, :attributes, :children, :type

    def prefixed_attributes
      attributes.inject({}) do |memo, (key, value)|
        memo[prefixed_attribute_name("@#{key}")] = value
        memo
      end
    end

    def prefixed_attribute_name(attribute)
      return attribute unless @options[:convert_tags_to].respond_to? :call
      @options[:convert_tags_to].call(attribute)
    end

    def add_node(node)
      @text = true if node.is_a? String
      @children << node
    end

    def to_hash
      if @type == "file"
        f = StringIOFile.new((@children.first || '').unpack('m').first)
        f.original_filename = attributes['name'] || 'untitled'
        f.content_type = attributes['content_type'] || 'application/octet-stream'
        return { name => f }
      end

      if @text
        t = typecast_value(inner_html)
        t = advanced_typecasting(t) if t.is_a?(String) && @options[:advanced_typecasting]

        if t.is_a?(String)
          t = StringWithAttributes.new(t)
          t.attributes = attributes
        end

        return { name => t }
      else
        #change repeating groups into an array
        groups = @children.inject({}) { |s,e| (s[e.name] ||= []) << e; s }

        out = nil
        if @type == "array"
          out = []
          groups.each do |k, v|
            if v.size == 1
              out << v.first.to_hash.entries.first.last
            else
              out << v.map{|e| e.to_hash[k]}
            end
          end
          out = out.flatten

        else # If Hash
          out = {}
          groups.each do |k,v|
            if v.size == 1
              out.merge!(v.first)
            else
              out.merge!( k => v.map{|e| e.to_hash[k]})
            end
          end
          out.merge! prefixed_attributes unless attributes.empty?
          out = out.empty? ? nil : out
        end

        if @type && out.nil?
          { name => typecast_value(out) }
        else
          { name => out }
        end
      end
    end

    # Typecasts a value based upon its type. For instance, if
    # +node+ has #type == "integer",
    # {{[node.typecast_value("12") #=> 12]}}
    #
    # @param value<String> The value that is being typecast.
    #
    # @details [:type options]
    #   "integer"::
    #     converts +value+ to an integer with #to_i
    #   "boolean"::
    #     checks whether +value+, after removing spaces, is the literal
    #     "true"
    #   "datetime"::
    #     Parses +value+ using Time.parse, and returns a UTC Time
    #   "date"::
    #     Parses +value+ using Date.parse
    #
    # @return <Integer, TrueClass, FalseClass, Time, Date, Object>
    #   The result of typecasting +value+.
    #
    # @note
    #   If +self+ does not have a "type" key, or if it's not one of the
    #   options specified above, the raw +value+ will be returned.
    def typecast_value(value)
      return value unless @type
      proc = self.class.typecasts[@type]
      proc.nil? ? value : proc.call(value)
    end

    def advanced_typecasting(value)
      split = value.split
      return value if split.size > 1

      case split.first
        when "true"       then true
        when "false"      then false
        when XS_DATE_TIME then try_to_convert(value) {|x| DateTime.parse(x)}
        when XS_DATE      then try_to_convert(value) {|x| Date.parse(x)}
        when XS_TIME      then try_to_convert(value) {|x| Time.parse(x)}
        else                   value
      end
    end

    # Take keys of the form foo-bar and convert them to foo_bar
    def undasherize_keys(params)
      params.keys.each do |key, value|
        params[key.tr("-", "_")] = params.delete(key)
      end
      params
    end

    # Get the inner_html of the REXML node.
    def inner_html
      @children.join
    end

    # Converts the node into a readable HTML node.
    #
    # @return <String> The HTML node in text form.
    def to_html
      attributes.merge!(:type => @type ) if @type
      "<#{name}#{attributes.to_xml_attributes}>#{@nil_element ? '' : inner_html}</#{name}>"
    end
    alias to_s to_html

  private
    def try_to_convert(value, &block)
      block.call(value)
    rescue ArgumentError
      value
    end

    def strip_namespace(string)
      string.split(":").last
    end
  end
end
